src/red_black_tree.rs

```diff
@@ -795,3 +929,282 @@ impl<
+    assert_eq!(tree.get_parent(grandparent), parent);
+    assert_eq!(tree.get_parent(leaf), parent);
+    assert!(tree.is_leaf(leaf) && tree.is_leaf(grandparent));
+}
+
+#[test]
+/// This test addresses the case where a node's parent is red and uncle is black.
+/// The new leaf is the right child of the parent and the parent is the left
+/// child of the grandparent.
+///
+/// "P is left child of G and L is right child of P."
+///
+/// We resolve this by rotating the parent left then applying the same
+/// algorithm as the previous test.
+fn test_right_insert_with_red_left_child_parent_and_black_uncle() {
+    type Rbt = RedBlackTree<u64, u64, 1024>;
+    let mut buf = vec![0u8; std::mem::size_of::<Rbt>()];
+    let tree = Rbt::new_from_slice(buf.as_mut_slice());
+    let addrs = vec![
+        tree.insert(61, 0).unwrap(),
+        tree.insert(85, 0).unwrap(),
+        tree.insert(52, 0).unwrap(),
+        tree.insert(41, 0).unwrap(),
+    ];
+
+    let parent = addrs[3];
+    // Uncle is black as it is null
+    let grandparent = addrs[2];
+
+    assert!(tree.is_black(addrs[0]) && tree.is_black(addrs[1]) && tree.is_black(grandparent));
+    assert!(tree.is_red(parent));
+
+    assert_eq!(tree.get_right(addrs[0]), addrs[1]);
+    assert_eq!(tree.get_left(addrs[0]), grandparent);
+    assert_eq!(tree.get_parent(addrs[1]), addrs[0]);
+    assert_eq!(tree.get_parent(grandparent), addrs[0]);
+
+    assert_eq!(tree.get_right(grandparent), SENTINEL);
+    assert_eq!(tree.get_left(grandparent), parent);
+    assert_eq!(tree.get_parent(parent), grandparent);
+
+    let leaf = tree.insert(47, 0).unwrap();
+
+    assert!(tree.is_black(addrs[0]) && tree.is_black(addrs[1]) && tree.is_black(leaf));
+    assert!(tree.is_red(grandparent) && tree.is_red(parent));
+
+    assert_eq!(tree.get_right(addrs[0]), addrs[1]);
+    assert_eq!(tree.get_left(addrs[0]), leaf);
+    assert_eq!(tree.get_parent(addrs[1]), addrs[0]);
+    assert_eq!(tree.get_parent(leaf), addrs[0]);
+
+    assert_eq!(tree.get_right(leaf), grandparent);
+    assert_eq!(tree.get_left(leaf), parent);
+    assert_eq!(tree.get_parent(grandparent), leaf);
+    assert_eq!(tree.get_parent(parent), leaf);
+    assert!(tree.is_leaf(parent) && tree.is_leaf(grandparent));
+}
+
+#[test]
+fn test_delete_multiple_random() {
+    use std::collections::hash_map::DefaultHasher;
+    use std::hash::{Hash, Hasher};
+    type Rbt = RedBlackTree<u64, u64, 1024>;
+    let mut buf = vec![0u8; std::mem::size_of::<Rbt>()];
+    let tree = Rbt::new_from_slice(buf.as_mut_slice());
+    let mut keys = vec![];
+    // Fill up tree
+    for k in 0..1024 {
+        let mut hasher = DefaultHasher::new();
+        (k as u64).hash(&mut hasher);
+        let key = hasher.finish();
+        tree.insert(key, 0).unwrap();
+        keys.push(key);
+        assert!(tree.is_valid_red_black_tree());
+    }
+
+    for i in keys.iter() {
+        tree.remove(i).unwrap();
+        assert!(tree.is_valid_red_black_tree());
```

Comment:
This line should not be trivial to pass, the fact that it does gives me confidence that the code is doing what it needs to be doing